drivers/char/broadcom/vc_mem.c

```diff
@@ -157,6 +164,115 @@ vc_mem_compat_ioctl(struct file *file, unsigned int cmd, unsigned long arg)
 }
 #endif
 
+static loff_t
+vc_mem_llseek(struct file *file, loff_t off, int whence) {
+	switch (whence) {
+		case SEEK_CUR:
+			off += file->f_pos;
+			// fallthrough
+		case SEEK_SET:
+			file->f_pos = off;
+			break;
+		default:
+			return -EINVAL;
+	}
+
+	force_successful_syscall_return();
+	return off;
+}
+
+static ssize_t
+vc_mem_read(struct file *file, char __user *read_buf, size_t read_size, loff_t *ppos) {
+	int err = 0;
+	void *buf;
+	dma_addr_t buf_paddr;
+	dma_addr_t src_begin, src_end, src_at;
+
+	struct vc_mem *drv = file->private_data;
+	struct device *dev = drv->misc.parent;
+	struct dma_chan *dma_chan = drv->dma_chan;
+	struct dma_device *dma_dev = dma_chan->device;
+
+	unsigned long init_pos = *ppos;
+	const unsigned long alias_cached = 0x00000000;
+	const unsigned long alias_direct = 0xc0000000;
+	const unsigned long alias_mask = 0xc0000000;
+
+	// Check the alias bits are permitted
+	if ((init_pos & alias_mask) != alias_cached &&
+	    (init_pos & alias_mask) != alias_direct &&
+	    (init_pos & alias_mask) != (drv->phys_addr & alias_mask))
+		return -EINVAL;
+
+	// Discard the cache alias bits for bounds checking
+	init_pos &= ~alias_mask;
+
+	if (init_pos < drv->base || init_pos > drv->size ||
+	    init_pos + read_size > drv->size)
```

Comment:
On 32-bit we are relying on the fact that `read_size <= MAX_RW_COUNT == 0x7ffff000` and on the `&= ~alias_mask` above. If either were not the case, `init_pos + read_size` could overflow.

I think this should either be changed to something that never overflows or explained with a comment.